### DISCUSSION TITLE
Make 'cropfactor' a property of calibration, not lens

### DISCUSSION THREAD

**Discussion by gnl21** [lensfun/lensfun#1765] [Ideas]
Currently the database has 'cropfactor' as a property of lenses, but the value really represents the crop factor that was used to create the calibration data. This means that the database contains multiple copies of the naming for lenses with multiple calibrations at different crop factors. If the cropfactor were made a property of the calibration then this could be handled more directly. I think it would make the database structure a little bit nicer, and it would allow lensfun to do a better job of handling lenses with multiple calibrations.

My motivation for making this change is that one of the lenses that I have (Canon EF50 f/1.8 II, although it's not really relevant), has calibrations for crop factors 1.0, 1.3 and 1.6. I provided the calibration for crop factor 1.6 with distortion, TCA and vignetting, but lensfun prefers the calibration for crop factor 1.3 (apparently the distortion correction is a little more accurate when profiled on a larger sensor). At that point there's really no good choices for a user to get corrections for their lens -- even profiling it and contributing a profile hasn't worked. Since the different calibrations are viewed internally as totally different lenses, there's not much that can be done in the code either. By getting a better view of the calibrations that are associated with the lens in question, we could make better decisions about what data to return. In this example, either the 1.6 crop calibration could be returned, or the data could be combined to get the best of both calibrations. (The 1.6 crop distortion calibration is also unused in lensfun at the moment, since the 1.3 crop version is more accurate, so we could remove it from the database. In fact, perhaps all we need is the 1.0 crop calibration, since the others can all be derived from it, but I don't know what effect that would have on accuracy).

This would require a change in the database schema, but I think it would be worth considering for a future version.

**Comment by sarunasb** [comment]
If I understand correctly, it would be something like this for a lens, which has calibration data on, for example, "full-frame", Nikon APS-C and FourThirds/microFourThirds?:
```
<lens>
        <maker>Nikon</maker>
        <model>NIKKOR Z 14-24mm f/2.8 S</model>
        <mount>Nikon Z</mount>
        <calibration cropfactor="1">
            (correction params)
        </calibration>
        <calibration cropfactor="1.5">
            (correction params)
        <calibration>
        <calibration cropfactor="2" aspect-ratio="4:3">
            (correction params)
        <calibration>
</lens>
```
On the other hand, having correction data for a bigger sensor, wouldn't it be possible to apply it to a smaller sensor, which fits completely within the area of bigger one? In such a case `<calibration>` sections for any smaller sensor wouldn't be even needed...

**Reply by gnl21** [reply]
Yes, that's how I imagined it would look.

I touched on the idea of just using the calibration for the larger sensor at the end of my comment. I don't know whether you'd get enough accuracy over the smaller area with the curve fitted over the larger area. (I'm also not sure if our calibration lets us get anywhere near that accuracy anyway).

Even if we can apply the larger calibration to the smaller sensor, I think we'd still want to support having multiple calibrations at different crop factors. For example, if a lens has a calibration for distortion only with crop factor 1, we would want to allow someone with a crop camera to add tca and vignetting. That would require this change, and once you've done that, it's probably easiest to make the change with the existing data and then worry about removing calibrations that turn out to be redundant afterwards.
